tests/Feature/GroupBalancedDivisionTest.php

```diff
@@ -0,0 +1,148 @@
+use App\Models\Event;
+use App\Models\User;
+use App\Models\Course;
+use App\Models\Registration;
+use App\Models\Group;
+
+
+uses(\Illuminate\Foundation\Testing\RefreshDatabase::class);
+
+test('equal div only drinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    foreach ($courses as $course) {
+        User::factory()->count(20)->create(['course_id' => $course->id]);
+    }
+
+    foreach (User::all() as $user) {
+        Registration::factory()->create([
+            'event_id' => $event->id,
+            'user_id' => $user->id,
+            'drinks_alcohol' => true,
+            'fulfils_requirements' => true,
+            'queue_position' => null
+        ]);
+    }
+
+    $div = new GroupBalancedDivision($event, true);
+    $div->assign();
+
+    // each group should contain 2 users per course
+    foreach ($groups as $group) {
+        foreach ($courses as $course) {
+            expect($group->registrations()
+                ->join('users', 'registrations.user_id', '=', 'users.id')
+                ->where('users.course_id', '=', $course->id)
+                ->count())->toEqual(2);
+        }
+    }
+});
+
+test('unequal div only drinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    $i = 10;
+
+    foreach ($courses as $course) {
+        User::factory()->count($i)->create(['course_id' => $course->id]);
+        $i += 10;
+    }
+
+    foreach (User::all() as $user) {
+        Registration::factory()->create([
+            'event_id' => $event->id,
+            'user_id' => $user->id,
+            'drinks_alcohol' => true,
+            'fulfils_requirements' => true,
+            'queue_position' => null
+        ]);
+    }
+
+    $div = new GroupBalancedDivision($event, true);
+    $div->assign();
+
+    foreach ($groups as $group) {
+        $i = 1;
+        foreach ($courses as $course) {
+            expect($group->registrations()
+                ->join('users', 'registrations.user_id', '=', 'users.id')
+                ->where('users.course_id', '=', $course->id)
+                ->count())->toEqual($i);
+            $i += 1;
+        }
+    }
+});
+
+test('unequal div nondrinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    $i = 10;
+
+    foreach ($courses as $course) {
+        User::factory()->count($i)->create(['course_id' => $course->id]);
+        $i += 10;
+    }
+
+    $checks = 0;
+    $balance = 0.0;
+
+    for ($n = 0; $n < 10; $n++) {
```

Comment:
What added value is expected from this implementation? The use of Magic Numbers in particular should not occur at an arbitrary point in the test. Instead, the whole thing should be structured in such a way that the entire case is executed several times.